pex/cli/commands/lock.py

```diff
@@ -119,6 +131,237 @@ def __call__(self, parser, namespace, value, option_str=None):
+
+    def sync(
+        self,
+        distributions,  # type: Iterable[Distribution]
+        confirm=True,  # type: bool
+        retain_pip=False,  # type: bool
+    ):
+        # type: (...) -> Result
+
+        abs_venv_dir = os.path.realpath(self.venv.venv_dir)
+
+        existing_distributions_by_project_name = {
+            dist.metadata.project_name: dist for dist in self.venv.iter_distributions()
+        }  # type: Dict[ProjectName, Distribution]
+
+        to_remove = []  # type: List[Distribution]
+        to_install = []  # type: List[Distribution]
+        for distribution in distributions:
+            existing_distribution = existing_distributions_by_project_name.pop(
+                distribution.metadata.project_name, None
+            )
+            if not existing_distribution:
+                to_install.append(distribution)
+            elif existing_distribution.metadata.version != distribution.metadata.version:
+                to_remove.append(existing_distribution)
+                to_install.append(distribution)
+        if retain_pip:
+            existing_distributions_by_project_name.pop(ProjectName("pip"), None)
+        to_remove.extend(existing_distributions_by_project_name.values())
+
+        to_unlink_by_pin = (
+            OrderedDict()
+        )  # type: OrderedDict[Tuple[ProjectName, Version], List[Text]]
+        for distribution in to_remove:
+            to_unlink = []  # type: List[Text]
+            if distribution.metadata.type is MetadataType.DIST_INFO:
+                to_unlink.extend(
+                    os.path.realpath(os.path.join(distribution.location, installed_file.path))
+                    for installed_file in Record.read(distribution.iter_metadata_lines("RECORD"))
+                )
+            elif distribution.metadata.type is MetadataType.EGG_INFO:
+                installed_files = distribution.metadata.files.metadata_file_rel_path(
+                    "installed-files.txt"
+                )
+                if installed_files:
+                    base_dir = os.path.realpath(
+                        os.path.join(distribution.location, os.path.dirname(installed_files))
+                    )
+                    to_unlink.extend(
+                        os.path.join(base_dir, path)
+                        for path in distribution.iter_metadata_lines("installed-files.txt")
+                    )
+                    # Pip generates "installed-files.txt" upon installing and this file does not
+                    # include itself; so we tack that on here ourselves.
+                    to_unlink.append(
+                        os.path.realpath(os.path.join(distribution.location, installed_files))
+                    )
+            if to_unlink:
+                to_unlink_by_pin[
+                    (distribution.metadata.project_name, distribution.metadata.version)
+                ] = [file for file in to_unlink if abs_venv_dir == commonpath((abs_venv_dir, file))]
```

Comment:
I've noticed this idiom is used in a quite a few places (not just this PR), is it worth a dedicated `is_prefix_of` (or something) function for it?

Reply:
I guess I haven't thought so. The `commonpath` function is stdlib and this is about the only way to use it; so why obscure a stdlib idiom behind a bespoke to the Pex codebase helper.